Documentation/anti-patterns.md

```diff
@@ -0,0 +1,180 @@
+## Anti-Patterns
+
+### Dialing in gRPC
+`grpc.Dial` is a function in the gRPC library that creates a virtual connection
+from the gRPC client to the gRPC server.  It takes a target URI (which can
+represent the name of a logical backend service and could resolve to multiple
+actual addresses) and a list of options, and returns a `ClientConn` object that
+represents the connection to the server. The `ClientConn` contains one or more
+actual connections to real server backends and attempts to keep these
+connections healthy by automatically reconnecting to them when they break.
+
+The `Dial` function can also be configured with various options to
+customize the behavior of the client connection. For example, developers could
+use options such a `WithTransportCredentials` to configure the transport
+credentials to use.
+
+While `Dial` is commonly referred to as a "dialing" function, it doesn't
+actually perform the low-level network dialing operation like `net.Dial` would.
+Instead, it creates a virtual connection from the gRPC
+client to the gRPC server.
+
+`Dial` does initiate the process of connecting to the server, but it uses
+the ClientConn object to manage and maintain that connection over time. This is
+why errors encountered during the initial connection are no different from those
+that occur later on, and why it's important to handle errors from RPCs rather
+than relying on options like `FailOnNonTempDialError`, `WithBlock`, and
+`WithReturnConnectionError`.
+In fact, `Dial` does not always establish a connection to servers by default. The connection
+behavior is determined by the load balancing policy being used. For instance, an "active"
+load balancing policy such as Round Robin attempts to maintain a constant connection, while the default "pick first"
+policy delays connection until an RPC is executed. To explicitly initiate a connection, you can employ
+the WithBlock option.
+
+### Using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`
+
+The gRPC API provides several options that can be used to configure the behavior
+of dialing and connecting to a gRPC server. Some of these options, such as
+`FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`, rely on
+failures at dial time. However, we strongly discourage developers from using
+these options, as they can introduce race conditions and result in unreliable
+and difficult-to-debug code.
+
+###  Why we discourage using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`
```

Comment:
To me, the biggest reason, which is not even mentioned, is that connections can die at any point in time.  Which means you need to deal with RPC failures caused by this no matter whether a connection was never created in the first place, or whether it was created and then immediately lost.